imap-dl support for Kerberos/GSSAPI (and ssl_ciphers)

frozencemetery opened the pull request:
Refactoring commit is separate to keep the diff readable; it's fine to squash if you prefer.

This adds an optional dependency on [python3-gssapi](https://github.com/pythongssapi/python-gssapi/).  In full disclosure, I am an author of that library; I'm also the author of the GSSAPI support in offlineimap (which is why the code probably looks similar).  Unfortunately we don't have type annotations; I've looked into adding them, but using mypy with cython code doesn't seem to work (yet?).

Tested against a Kerberized Zimbra instance.  `ssl_ciphers` support is unfortunately needed for my local site, so it's first in the series.

Thanks!

frozencemetery commented:
At @dkg's suggestion, added more motivating context to the ssl_ciphers commit since I'm not aware of public servers displaying this problem.

dkg commented on imap-dl:
```diff
@@ -99,6 +105,52 @@ def auth_builtin(username:str, imap:imaplib.IMAP4_SSL,
+            token = None
+        if not self.gss_vc.complete:
+            response = self.gss_vc.step(token)
+            return response if response else b""
```
`mypy --strict` says:
> Returning Any from function declared to return "bytes"

dkg commented on imap-dl:
```diff
@@ -99,6 +105,52 @@ def auth_builtin(username:str, imap:imaplib.IMAP4_SSL,
+            reply[5:] = self.username.encode("utf-8")
+
+        response = self.gss_vc.wrap(bytes(reply), response.encrypted)
+        return response.message if response.message else b""
```
`mypy --strict` says:
> Returning Any from function declared to return "bytes"

frozencemetery commented:
Rebased onto your branch from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=953582 and added documentation of the new options.

gssapi doesn't have mypy annotations (yet - I'm working on it), so I've just squashed the mypy warnings.

dkg commented:
https://bugs.debian.org/953582 (and https://salsa.debian.org/dkg/mailscripts/-/tree/documentation) have an update to the imap-dl documentation that might make it easier to land documentation patches.

btw, you might also want to update the debian/control dependencies to recommend (or suggest) `python3-gssapi`.

dkg commented:
(our last two messages must have crossed in the ether, sorry for the out-of-order notes above)

Thanks for these updates, @frozencemetery !  if your commit message that introduces the `#type: ignore" markers to quiet the mypy warnings can mention why they're there, that'll help anyone in the future who is trying to clean up the type annotations to know what's going on.  (eventually we want to remove the `#type: ignore`, right?)

dkg commented on imap-dl.1.pod:
```diff
@@ -48,6 +48,10 @@ B<retriever.username> is the username of the IMAP account.
 plaintext passwords.
 
+B<retriever.ssl_ciphers> is an OpenSSL cipher string to use instead of the
+defaults.  (The defaults are good; this should be avoided except to work aroun
```
s/aroun/around/

dkg commented on imap-dl.1.pod:
```diff
@@ -48,6 +48,10 @@ B<retriever.username> is the username of the IMAP account.
 B<retriever.password> is the password for the IMAP account when using
 plaintext passwords.
 
+B<retriever.use_kerberos> requests that Kerberos (through GSSAPI) is to be
```
maybe indicate that this is a boolean?

frozencemetery commented:
Updated, addressing review comments.
